### ISSUE TITLE
Why do I get different results when running the same data multiple times?

### ISSUE THREAD

**Issue by woaiwojia4816294** [richzhang/PerceptualSimilarity#106]
import lpips
import os
from tqdm import tqdm
from PIL import Image
import numpy as np
import torch

loss_fn_alex = lpips.LPIPS(net='alex').cuda()

dataPath = r'D:\PycharmProjects\paper_instrument\ImageQualityAssessment\hdrnet'
groundTruthPath = r'D:\PycharmProjects\paper_instrument\ImageQualityAssessment\groundtruth'
assert len(os.listdir(dataPath)) == len(os.listdir(groundTruthPath))

LPIPS = 0.0
for x_name, y_name in tqdm(zip(os.listdir(dataPath), os.listdir(groundTruthPath))):
    x = (np.array(Image.open(os.path.join(dataPath, x_name))).transpose(2, 0, 1).astype(np.float32) / 255) * 2 - 1
    y = (np.array(Image.open(os.path.join(groundTruthPath, y_name))).transpose(2, 0, 1).astype(np.float32) / 255) * 2 - 1
    x = torch.from_numpy(x).unsqueeze(0).cuda()
    y = torch.from_numpy(y).unsqueeze(0).cuda()
    LPIPS += loss_fn_alex(x, y)

avgLPIPS = LPIPS / len(os.listdir(dataPath))
print(avgLPIPS)

**Comment by richzhang** [comment]
Double check that it's in eval by setting `.eval()`

**Comment by visittor** [comment]
I also got the same behaviour even after I created loss_fn_alex with use_dropout=False and called loss_fn_alex.eval.

**Comment by QingXin96** [comment]
I also got very strange results, I had 7 pair of images for evaluation, and the result each time is not the same,  moreover, it seems a permute of a fixed array. I also tried in case of disenabling dropout and enable .eval() but got the same behavior.

First test:      0.0115,  0.0115, 0.0059, 0.0205, 0.0204, 0.0268, 0.0247
Second test: 0.0205, 0.0059, 0.0268, 0.0115, 0.0115, 0.0247, 0.0204
Third  test:    0.0115, 0.0247, 0.0204, 0.0268, 0.0205, 0.0115, 0.0059

Can you give a solution? since I was asked by the reviewers to add this evaluation to my paper.

**Comment by byliutao** [comment]
I have the same issue with the latest torch version. I find if I set lpips=False, this result will be the same.

**Comment by byliutao** [comment]
i find this acctually because the lpips weight don't load successfully. The .pth file's weights have a different keys value compare to the lpips network itself, this may due the the torch version's problem(I use the latest torch version). 

you can use below FIX_LPIPS class to replace original LPIPS class to load lpips weight successfully.
```
class FIX_LPIPS(LPIPS):
    def __init__(self, pretrained=True, net='alex', version='0.1', lpips=True, spatial=False, 
        pnet_rand=False, pnet_tune=False, use_dropout=False, model_path=None, eval_mode=True, verbose=True):
        """ Initializes a perceptual loss torch.nn.Module

        Parameters (default listed first)
        ---------------------------------
        lpips : bool
            [True] use linear layers on top of base/trunk network
            [False] means no linear layers; each layer is averaged together
        pretrained : bool
            This flag controls the linear layers, which are only in effect when lpips=True above
            [True] means linear layers are calibrated with human perceptual judgments
            [False] means linear layers are randomly initialized
        pnet_rand : bool
            [False] means trunk loaded with ImageNet classification weights
            [True] means randomly initialized trunk
        net : str
            ['alex','vgg','squeeze'] are the base/trunk networks available
        version : str
            ['v0.1'] is the default and latest
            ['v0.0'] contained a normalization bug; corresponds to old arxiv v1 (https://arxiv.org/abs/1801.03924v1)
        model_path : 'str'
            [None] is default and loads the pretrained weights from paper https://arxiv.org/abs/1801.03924v1

        The following parameters should only be changed if training the network

        eval_mode : bool
            [True] is for test mode (default)
            [False] is for training mode
        pnet_tune
            [False] tune the base/trunk network
            [True] keep base/trunk frozen
        use_dropout : bool
            [True] to use dropout when training linear layers
            [False] for no dropout when training linear layers
        """

        super(LPIPS, self).__init__()
        if(verbose):
            print('Setting up [%s] perceptual loss: trunk [%s], v[%s], spatial [%s]'%
                ('LPIPS' if lpips else 'baseline', net, version, 'on' if spatial else 'off'))

        self.pnet_type = net
        self.pnet_tune = pnet_tune
        self.pnet_rand = pnet_rand
        self.spatial = spatial
        self.lpips = lpips # false means baseline of just averaging all layers
        self.version = version
        self.scaling_layer = ScalingLayer()

        if(self.pnet_type in ['vgg','vgg16']):
            net_type = pn.vgg16
            self.chns = [64,128,256,512,512]
        elif(self.pnet_type=='alex'):
            net_type = pn.alexnet
            self.chns = [64,192,384,256,256]
        elif(self.pnet_type=='squeeze'):
            net_type = pn.squeezenet
            self.chns = [64,128,256,384,384,512,512]
        self.L = len(self.chns)

        self.net = net_type(pretrained=not self.pnet_rand, requires_grad=self.pnet_tune)

        if(lpips):
            self.lin0 = NetLinLayer(self.chns[0], use_dropout=use_dropout)
            self.lin1 = NetLinLayer(self.chns[1], use_dropout=use_dropout)
            self.lin2 = NetLinLayer(self.chns[2], use_dropout=use_dropout)
            self.lin3 = NetLinLayer(self.chns[3], use_dropout=use_dropout)
            self.lin4 = NetLinLayer(self.chns[4], use_dropout=use_dropout)
            self.lins = [self.lin0,self.lin1,self.lin2,self.lin3,self.lin4]
            if(self.pnet_type=='squeeze'): # 7 layers for squeezenet
                self.lin5 = NetLinLayer(self.chns[5], use_dropout=use_dropout)
                self.lin6 = NetLinLayer(self.chns[6], use_dropout=use_dropout)
                self.lins+=[self.lin5,self.lin6]
            self.lins = nn.ModuleList(self.lins)

            # 假设权重文件中的层名是 lin3.model.1.weight，需映射到代码中的 lin3.model.0.weight
            weight_map1 = {
                'lin0.model.1.weight': 'lin0.model.0.weight',
                'lin1.model.1.weight': 'lin1.model.0.weight',
                'lin2.model.1.weight': 'lin2.model.0.weight',
                'lin3.model.1.weight': 'lin3.model.0.weight',
                'lin4.model.1.weight': 'lin4.model.0.weight',             
            }      


            if(pretrained):
                if(model_path is None):
                    import inspect
                    import os
                    model_path = os.path.abspath(os.path.join(inspect.getfile(self.__init__), '..', 'weights/v%s/%s.pth'%(version,net)))

                if(verbose):
                    print('Loading model from: %s'%model_path)
                
                state_dict = torch.load(model_path, map_location='cpu')  # 加载原始权重
                mapped_state_dict1 = {weight_map1[k]: v for k, v in state_dict.items()}
                self.load_state_dict(mapped_state_dict1, strict=False)  # strict=False允许部分匹配
                # self.load_state_dict(torch.load(model_path, map_location='cpu'), strict=True)          

        if(eval_mode):
            self.eval()
```
